internal/controller/dnsrecord_accessor.go

```diff
@@ -0,0 +1,131 @@
+/*
+Copyright 2025.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	externaldns "sigs.k8s.io/external-dns/endpoint"
+
+	"github.com/kuadrant/dns-operator/api/v1alpha1"
+)
+
+var _ DNSRecordAccessor = &RemoteDNSRecord{}
+
+type DNSRecordAccessor interface {
+	metav1.Object
+	runtime.Object
+	v1alpha1.ProviderAccessor
+	GetDNSRecord() *v1alpha1.DNSRecord
+	GetOwnerID() string
+	GetRootHost() string
+	GetZoneDomainName() string
+	GetZoneID() string
+	GetSpec() *v1alpha1.DNSRecordSpec
+	GetStatus() *v1alpha1.DNSRecordStatus
+	SetStatusCondition(conditionType string, status metav1.ConditionStatus, reason, message string)
+	SetStatusZoneID(id string)
+	SetStatusZoneDomainName(domainName string)
+	SetStatusDomainOwners(owners []string)
+	SetStatusEndpoints(endpoints []*externaldns.Endpoint)
+	SetStatusObservedGeneration(observedGeneration int64)
+	HasOwnerIDAssigned() bool
+	HasDNSZoneAssigned() bool
+}
+
+type RemoteDNSRecord struct {
+	*v1alpha1.DNSRecord
+	ClusterID string
+	status    *v1alpha1.DNSRecordStatus
+}
+
+func (s *RemoteDNSRecord) GetDNSRecord() *v1alpha1.DNSRecord {
+	return s.DNSRecord
+}
+
+func (s *RemoteDNSRecord) GetOwnerID() string {
+	return s.DNSRecord.Status.OwnerID
+}
+
+func (s *RemoteDNSRecord) GetZoneDomainName() string {
+	return s.GetStatus().ZoneDomainName
+}
+
+func (s *RemoteDNSRecord) GetZoneID() string {
+	return s.GetStatus().ZoneID
+}
+
+func (s *RemoteDNSRecord) GetSpec() *v1alpha1.DNSRecordSpec {
+	return &s.Spec
+}
+
+// GetStatus returns the status set for the current cluster ID.
+// If none is set an empty DNSRecordStatus is returned.
+func (s *RemoteDNSRecord) GetStatus() *v1alpha1.DNSRecordStatus {
+	if s.status == nil {
```

Comment:
this reads a bit odd. So the cluster status is only set once ?

Reply:
It currently means that `setStatus()` must be called before doing an `Update` of the resource through the k8s API which I'm not totally happy with, and is the reason that all the `SetStatus*` methods are added to the accessor interface as they currently make the call to `setStatus()` on all updates to make sure it is actually set in the raw json.